src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala

```diff
@@ -15,21 +15,107 @@ import scala.tools.refactoring.util.SourceWithMarker.Movement
 import scala.tools.refactoring.common.TracingHelpers
 import scala.tools.refactoring.util.SourceHelpers
 import scala.tools.refactoring.util.SourceWithSelection
+import scala.tools.refactoring.util.Casts._
 
 trait MarkOccurrences extends common.Selections with analysis.Indexes with common.CompilerAccess with common.EnrichedTrees with common.InteractiveScalaCompiler {
   import global._
   import TracingHelpers.toCompactString
 
+  private def tryFindMissingSymbol(treeWithoutSymbol: Tree, root: Tree): Symbol = {
+    treeWithoutSymbol match {
+      case treeWithoutSymbol: RefTree =>
+        tryFindMissingSymbol(treeWithoutSymbol: RefTree, root: Tree)
+
+      case _ =>
+        NoSymbol
+    }
+  }
+
+  /*
+   * In some cases (for example when dealing with refined self types), the selected tree might not
+   * have a symbol. This method attempts to find it anyway.
+   */
+  private def tryFindMissingSymbol(treeWithoutSymbol: RefTree, root: Tree): Symbol = {
+    root.foreach { t =>
+      t.tryMatch { case t: TypeTree =>
+        if (t.original == treeWithoutSymbol) {
+          t.tpe.parents.foreach { p =>
+            p.tryMatch {
+              case p: TypeRef if p.sym.name == treeWithoutSymbol.name =>
+                  return p.sym
+            }
+          }
+        } else {
+          t.original.tryMatch {
+            case orig: RefTree if orig.qualifier == treeWithoutSymbol =>
+              t.tpe.parents.foreach { p =>
+                p.tryMatch { case p: TypeRef =>
+                  p.pre.tryMatch {
+                    case pre: ThisType if pre.sym.name.toString == treeWithoutSymbol.name.toString =>
+                      return pre.sym
+                  }
+                }
+              }
+
+            case orig: CompoundTypeTree =>
+              val tParents = t.tpe.parents.flatMap {
+                case t: RefinedType => t.parents
+                case _ => Nil
+              }
+
+              val oParents = orig.templ.parents
+
+              if (oParents.size == tParents.size) {
+                oParents.zip(tParents).foreach { case (oTree, tRef) =>
+                  oTree.tryMatch {
+                    case refTree: RefTree =>
+                      val sym = findSymbolInType(treeWithoutSymbol, tRef, refTree)
+                      if (sym != NoSymbol) {
+                        return sym
+                      }
+                  }
+                }
+              }
+          }
+        }
+      }
+    }
+
+    NoSymbol
+  }
+
+  private def findSymbolInType(treeWithoutSymbol: Tree, tpe: Type, ref: RefTree): Symbol = {
```

Comment:
What do you think about this:
```scala
def findSymbolInType(tree: Tree, ref: RefTree): Type => Symbol = {
  case ... =>
}
```
I understand that call is like `findSymbolInType(tree, ref)(tpe)` but maybe is better?